tests/token-guard.ts

```diff
@@ -381,4 +365,169 @@ describe("token-guard", () => {
+
+    it("initialises a tokenGuard with a membership token requirement", async () => {
+      membershipTokenMint = await Token.createMint(
+        provider.connection,
+        membershipTokenMinter,
+        membershipTokenMinter.publicKey,
+        membershipTokenMinter.publicKey,
+        0,
+        TOKEN_PROGRAM_ID
+      );
+
+      console.log(`Minted membership token ${membershipTokenMint.publicKey}`);
+
+      tokenGuardState = await initialize(
+        program,
+        provider,
+        gatekeeperNetwork.publicKey,
+        recipient.publicKey,
+        undefined,
+        undefined,
+        undefined,
+        {
+          key: membershipTokenMint.publicKey,
+          strategy: "SPL",
+        }
+      );
+    });
+
+    it("should not let someone without the membership token exchange", async () => {
+      const instructions = await exchange(
+        provider.connection,
+        program,
+        tokenGuardState.id,
+        sender.publicKey,
+        sender.publicKey,
+        gatekeeperNetwork.publicKey,
+        exchangeAmount
+      );
+
+      // fail, because the membership token account is not presented
+      const shouldFail = sendTransactionFromSender(instructions);
+
+      return expect(shouldFail).to.be.rejectedWith(
+        /Transaction simulation failed/
+      );
+    });
+
+    it("should not let someone with an insufficient balance of the membership token exchange", async () => {
+      senderMembershipTokenATA =
+        await membershipTokenMint.createAssociatedTokenAccount(
+          sender.publicKey
+        );
+      const instructions = await exchange(
+        provider.connection,
+        program,
+        tokenGuardState.id,
+        sender.publicKey,
+        sender.publicKey,
+        gatekeeperNetwork.publicKey,
+        exchangeAmount,
+        {
+          tokenAccount: senderMembershipTokenATA,
+        }
+      );
+
+      // fail, because the account is present, but it is empty
+      const shouldFail = sendTransactionFromSender(instructions);
+
+      return expect(shouldFail).to.be.rejectedWith(
+        /Transaction simulation failed/
+      );
+    });
+
+    it("should not let someone that does not own the membership token exchange", async () => {
+      const someRandomMembershipTokenATA =
+        await membershipTokenMint.createAssociatedTokenAccount(
+          web3.Keypair.generate().publicKey
+        );
+      const instructions = await exchange(
+        provider.connection,
+        program,
+        tokenGuardState.id,
+        sender.publicKey,
+        sender.publicKey,
+        gatekeeperNetwork.publicKey,
+        exchangeAmount,
+        {
+          tokenAccount: someRandomMembershipTokenATA,
+        }
+      );
+
+      // fail, because the account is present, but has the wrong owner
+      const shouldFail = sendTransactionFromSender(instructions);
+
+      return expect(shouldFail).to.be.rejectedWith(
+        /Transaction simulation failed/
+      );
```

Comment:
This test could actually be failing because the account is empty (just like the test before) and not because the account is not owned by the sender, right? Maybe it's worth funding the account to make sure the only possible reason for it to fail is ownership? Ideally there would be a specific error message, though.

Reply:
True -Ill update it